source/request_response.c

```diff
@@ -220,6 +220,34 @@ int aws_http_request_add_header(struct aws_http_request *request, struct aws_htt
     return AWS_OP_ERR;
 }
 
+int aws_http_request_add_header_array(
+    struct aws_http_request *request,
+    struct aws_http_header *headers,
+    size_t num_headers) {
+
+    AWS_PRECONDITION(request);
+    AWS_PRECONDITION(headers);
+    AWS_PRECONDITION(num_headers > 0);
+
+    const size_t beginning_headers_size = aws_array_list_length(&request->headers);
+
+    for (size_t i = 0; i < num_headers; ++i) {
+        if (aws_http_request_add_header(request, headers[i])) {
+            goto error;
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+
+error:
+    /* Remove all headers we added */
+    for (size_t len = aws_array_list_length(&request->headers); len > beginning_headers_size; --len) {
```

Comment:
ooh, nice catch on the esoteric error handling ;)

Reply:
:+1: